[CDAP-9514] Multiple column directives in Data Prep (Swap and Merge directives)

tonybach opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-9514

tonybach committed:
[CDAP-9514] Multi-column selections in DataPrep + Swap and Merge columns directives
- Implements Clear All and Select All behaviors
- Adds checkbox behavior on table headers, and syncs behavior in table and in Columns tab
- Determines which directives work for which number of selected columns, and implementes multi-column Delete and Keep directives
- Implements Swap directive and Merge directive
- Changes dynamically labels of Drop & Keep directives depending on number of selected columns

ajainarayanan commented on cdap-ui/app/cdap/components/DataPrep/ColumnActionsDropdown/index.js:
```diff
@@ -40,39 +44,73 @@ export default class ColumnActionsDropdown extends Component {
         id: shortid.generate(),
-        tag: FilterDirective
+        tag: FilterDirective,
+        colSize: 1
```
May be `requiredColCount`? `colSize` signifies more on the size of the column whereas here we want minimum number of columns to be selected to apply the directive.

ajainarayanan commented on cdap-ui/app/cdap/components/DataPrep/Directives/DropColumn/index.js:
```diff
@@ -26,7 +27,7 @@ export default class DropColumnDirective extends Component {
 
   applyDirective() {
-    let column = this.props.column;
+    let column = this.props.column.toString();
```
If `props.column` comes in as array would this be a problem?

elfenheart replied:
Array.toString() would return comma separated string already I believe, so should be OK

ajainarayanan replied:
ah! TIL: array.toString() returns comma separated values as string 👍

ajainarayanan commented on cdap-ui/app/cdap/components/DataPrep/store/DataPrepActionCreator.js:
```diff
@@ -51,6 +51,7 @@ export function execute(addDirective, shouldReset) {
     MyDataPrepApi.execute(params, requestBody)
       .subscribe((res) => {
         observer.onNext(res);
+        console.log(res);
```
Remove?

ajainarayanan reviewed (commented):
Minor comments.

elfenheart commented on cdap-ui/app/cdap/components/DataPrep/ColumnActionsDropdown/index.js:
```diff
@@ -156,14 +206,27 @@ export default class ColumnActionsDropdown extends Component {
+                  if (this.state.selectedHeaders.length !== directive.colSize && directive.colSize !== 0) {
+                    disabled = true;
+                  }
+                } else if (directive.colSize == 2) {
```
`===`

elfenheart commented on cdap-ui/app/cdap/components/DataPrep/DataPrepSidePanel/DataPrepSidePanel.scss:
```diff
@@ -113,26 +114,45 @@ $tab-header-height: 37px;
+        padding-top: 5px;
+        padding-left: 10px;
+        padding-bottom: 5px;
+        padding-right: 0px;
```
just use the `padding` property to specify the 4 paddings

tonybach commented:
Addressed comments. Will add a commit for the design changes in the MergeColumns modal (https://issues.cask.co/browse/CDAP-9514?focusedCommentId=29393&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-29393)

elfenheart commented on cdap-ui/app/cdap/text/text-en.yaml:
```diff
@@ -233,6 +279,26 @@ features:
+          DASH: Dash
+          PERIOD: Period
+          PIPE: Pipe
+          UNDERSCORE: Underscore
```
alphabetical sorting?

tonybach replied:
Ok, only the `COMMA` and `COLON` were out of order right? Fixed that.

ajainarayanan commented on cdap-ui/app/cdap/components/DataPrep/ColumnActionsDropdown/ColumnActionsDropdown.scss:
```diff
@@ -58,6 +63,7 @@
       font-weight: 400;
       line-height: 2;
       cursor: initial;
+      box-shadow: 1px 1px #cccccc;
```
Might want to add a little blur 
```
box-shadow: 1px 1px 5px #cccccc;
```
to get some depth.

@bdmogal suggestion: Add boxshadow to the main menu too. If you think its not in the scope of this PR thats fine we will anyways revamp the look and feel of dataprep as part of another PR. we can handle it there.

tonybach replied:
@ajainarayanan Already added the box-shadow on the main menu, but accidentally applied it to the wrong location... It's the `box-shadow: 1px 1px 5px #cccccc;` that was added to `.popover-content` in the `DataPrepSidePanel`, will move it to the `.popover-content` over here.

Once that's done then the second level pop-over will have the same depth, so maybe we don't need to add it?

ajainarayanan replied:
We might need it for the second level popover menu too if its on top of the right panel (columns/directive tab).

ajainarayanan commented:
Talked to @tonybach @bdmogal Lea. Decided to have "Delete Column" if no columns have been selected and "Delete Selected Columns" if one or more columns have been selected. The same holds for "Keep Columns" and "Keep Selected Columns"

ajainarayanan commented on cdap-ui/app/cdap/components/DataPrep/Directives/MergeColumns/index.js:
```diff
@@ -0,0 +1,302 @@
+            onKeyPress={this.handleKeyPress}
+            onChange={this.handleNewColumnInputChange}
+            placeholder={T.translate(`${PREFIX}.newColumnPlaceholder`)}
+            ref={ref => this.newColumnInputBox = ref}
```
Can we use `autoFocus` instead of getting a ref and doing it manually?

ajainarayanan commented on cdap-ui/app/cdap/components/DataPrep/Directives/MergeColumns/index.js:
```diff
@@ -0,0 +1,302 @@
+  switchColumnOrder() {
+    if (this.state.newColumnInput === `${this.state.firstColumn}_${this.state.secondColumn}`) {
+      this.setState({
+        newColumnInput: `${this.state.secondColumn}_${this.state.firstColumn}`
```
Do we need two setStates? We could probably combine them using Object.assign and constructing the final object for setState.

ajainarayanan commented on cdap-ui/app/cdap/components/DataPrep/Directives/MergeColumns/index.js:
```diff
@@ -0,0 +1,302 @@
+          className="form-control"
+          value={this.state.customDelimiter}
+          onChange={this.handleCustomDelimiterChange}
+          ref={ref => this.customDelimiterRef = ref}
```
Same here too. We could probably use `autoFocus` here right?

ajainarayanan commented on cdap-ui/app/cdap/components/DataPrep/DataPrepSidePanel/DataPrepSidePanel.scss:
```diff
@@ -144,15 +161,19 @@ $tab-header-height: 37px;
-          &.expanded {
-            background-color: #dedede;
+          &.selected {
+            background-color: $element-selected-color;
```
Just noticed this for some colors we have them separated out in a variable and some are not. Any specific reasons?

tonybach replied:
Didn't extract out some colors like #999999 or #cccccc, since they were mostly used for borders, but have extracted them out now.

ajainarayanan commented on cdap-ui/app/cdap/components/DataPrep/Directives/MergeColumns/index.js:
```diff
@@ -0,0 +1,302 @@
+  renderDetail() {
+    if (!this.state.isOpen) { return null; }
+    let disabledCondition = this.state.newColumnInput.length === 0
+    || (this.state.selectedDelimiter.substr(0, 6) === 'CUSTOM'
```
Any specific reason why we are not doing a whole word comparison? I am seeing this 
```
<option value="CUSTOMDELIMITER">
  {T.translate(`${PREFIX}.Delimiters.CUSTOMDELIMITER`)}
</option>
```
wouldn't `this.state.selectedDelimiter === 'CUSTOMDELIMITER'` easier? Also noticing the same check being performed in multiple places. Is it possible to extract this out as a function and reuse that everywhere?

elfenheart commented on cdap-ui/app/cdap/components/DataPrep/DataPrepSidePanel/DataPrepSidePanel.scss:
```diff
@@ -193,3 +217,18 @@ $tab-header-height: 37px;
+
+      &:hover {
+        cursor: pointer;
+        background-color: #eeeeee;
```
extract color?

elfenheart commented on cdap-ui/app/cdap/components/DataPrep/store/index.js:
```diff
@@ -50,12 +53,19 @@ const dataprep = (state = defaultInitialState, action = defaultAction) => {
+      // after any directive, remove selected header(s) if they're no longer in
+      // the list of headers
+      selectedHeadersDiff.length > 0
+      && { selectedHeaders: difference(state.selectedHeaders, selectedHeadersDiff) }
```
```selectedHeaders: state.selectedHeaders.filter((head) => return action.payload.headers.indexOf(head) !== -1)```

elfenheart commented:
👍  LGTM

leacd commented:
LGTM

tonybach commented:
Build: https://builds.cask.co/browse/CDAP-DRC6370-15
